pkg/controller/volume/attachdetach/util/util_test.go

```diff
@@ -0,0 +1,316 @@
+					Name:      "kube-apiserver",
+					Namespace: "default",
+				},
+				Spec: v1.PodSpec{
+					Volumes: []v1.Volume{
+						{
+							VolumeSource: v1.VolumeSource{
+								HostPath: hostPath,
+								PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+									ClaimName: "claim-name",
+									ReadOnly: false,
+								},
+							},
+						},
+					},
+				},
+			},
+			enableCSIMigration:   true,
+			wantPersistentVolume: persistentVolumeExampeWithCSI,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			actualSpec, _ := CreateVolumeSpec(test.podVolume, test.pod, nodeName, plugMgr, pvcLister, pvLister, intreeToCSITranslator, csiTranslator)
+			if test.enableCSIMigration {
+				gotVolumeWithCSIMigration := *actualSpec.PersistentVolume
+				if gotVolumeWithCSIMigration.Name != test.wantPersistentVolume.Name {
+					t.Errorf("got volume name is %v, want volume name is %v", gotVolumeWithCSIMigration.Name, test.wantPersistentVolume.Name)
+
+				}
+				if !reflect.DeepEqual(gotVolumeWithCSIMigration.Spec, test.wantPersistentVolume.Spec) {
+					t.Errorf("got volume.Spec %v, want.Spec %v", gotVolumeWithCSIMigration, test.wantPersistentVolume)
+				}
+			} else {
+				gotVolume := *actualSpec.Volume
+				if !reflect.DeepEqual(gotVolume, test.wantVolume) {
+					t.Errorf("got volume %v, want %v", gotVolume, test.wantVolume)
+
+				}
+			}
+		})
+	}
+}
+
+func createObjects(pvMode, podMode v1.PersistentVolumeMode) (*v1.Node, *v1.Pod, *v1.PersistentVolume, *v1.PersistentVolumeClaim) {
+	node := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{Name: testHostName},
+		Status: v1.NodeStatus{
+			VolumesAttached: []v1.AttachedVolume{
+				{
+					Name:       "fake/fake-device",
+					DevicePath: "fake/path",
+				},
+			}},
+	}
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "abc",
+			Namespace: "default",
+			UID:       "1234",
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name: "container1",
+				},
+			},
+			Volumes: []v1.Volume{
+				{
+					Name: "vol1",
+					VolumeSource: v1.VolumeSource{
+						PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+							ClaimName: "claim-name",
+						},
+					},
+				},
+			},
+			SecurityContext: &v1.PodSecurityContext{
+				SupplementalGroups: []int64{555},
+			},
+		},
+	}
+	switch podMode {
+	case v1.PersistentVolumeBlock:
+		pod.Spec.Containers[0].VolumeDevices = []v1.VolumeDevice{
+			{
+				Name:       "vol1",
+				DevicePath: "/dev/vol1",
+			},
+		}
+	case v1.PersistentVolumeFilesystem:
+		pod.Spec.Containers[0].VolumeMounts = []v1.VolumeMount{
+			{
+				Name:      "vol1",
+				MountPath: "/mnt/vol1",
+			},
+		}
+	default:
+		// The volume is not mounted nor mapped
+	}
+	pv := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "pvA",
+		},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				RBD: &v1.RBDPersistentVolumeSource{
```

Comment:
Is this used anywhere? It looks like your test case examples are using hostpath and gce volume types.

Reply:
This is a random pv created for fake.NewSimpleClientset() set up.

Reply:
It looks like the function can take in an arbitrary set of objects: https://github.com/kubernetes/kubernetes/blob/1b141f021935910a9b937cc4e2469cbc8c098967/staging/src/k8s.io/client-go/kubernetes/fake/clientset_generated.go#L134

So I would only initialize it with objects you need.

Reply:
I removed the whole setup for now.